coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/functions/CoalesceStructUtility.java

```diff
@@ -0,0 +1,173 @@
+   * in its own format, Coral tries to maintain backward compatibility on top of that. Notably we also have
+   * Iceberg reader does the same, see Linkedin's (temporary) fork on Iceberg:
+   * https://github.com/linkedin/iceberg/pull/84 (Avro)
+   * https://github.com/linkedin/iceberg/pull/85 (ORC)
+   *
+   *
+   * Further details:
+   * struct_tr looks like:
+   * struct&lt;tag:int, field0:type0, field1:type1, ... fieldN:typeN&gt;
+   *
+   * struct_ex looks like:
+   * struct&lt;tag_0:type0, tag_1:type1, ... tag_N:typeN&gt;
+   *
+   * This new UDF could be stated as the following signatures:
+   * def coalesce_struct(struct:struct_tr) : struct_ex = {...}
+   * def coalesce_struct(struct:struct_tr, ordinal: int): field_at_ordinal = {...}
+   *
+   */
+  public static final SqlReturnTypeInference COALESCE_STRUCT_FUNCTION_RETURN_STRATEGY = opBinding -> {
+    int numArgs = opBinding.getOperandCount();
+    RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+    Preconditions.checkState(numArgs == 1 || numArgs == 2);
+    RelDataType coalescedDataType = coalesce(opBinding.getOperandType(0), typeFactory);
+    // 1-arg case
+    if (numArgs == 1) {
+      return coalescedDataType;
+    }
+    // 2-arg case
+    else {
+      int ordinal = opBinding.getOperandLiteralValue(1, Integer.class);
+      return coalescedDataType.getFieldList().get(ordinal).getType();
+    }
+  };
+  private static final String TRINO_PREFIX = "field";
+  private static final String HIVE_EXTRACT_UNION_PREFIX = "tag_";
+
+  private CoalesceStructUtility() {
+    // Utility class, does nothing in constructor
+  }
+
+  /**
+   * Converting a {@link RelDataType} that could potentially contains a Trino-format exploded-union(i.e. a struct
+   * in a format of {tag, field0, field1, ..., fieldN} to represent a union after being deserialized)
+   * into a exploded-union that complies with Hive's extract_union UDF format
+   * (i.e. a struct as {tag_0, tag_1, ..., tag_{N}} to represent a union after being deserialized)
+   *
+   * For more information, check: https://github.com/trinodb/trino/pull/3483
+   */
+  @VisibleForTesting
+  static RelDataType coalesce(RelDataType inputNode, RelDataTypeFactory typeFactory) {
+    // Using type information implicitly carried in the object of RelDateType
+    // instead of getting down to SqlTypeName since the former contains enough categorization
+    // of types to achieve the purpose for this method.
+
+    if (inputNode.isStruct()) {
+      List<String> fieldNames = inputNode.getFieldNames();
+      return coalesceStruct(inputNode, isTrinoStructPattern(fieldNames), typeFactory);
+    } else if (inputNode.getKeyType() != null) {
+      return coalesceMap(inputNode, typeFactory);
+    } else if (inputNode.getComponentType() != null) {
+      return coalesceCollection(inputNode, typeFactory);
+    } else {
+      return inputNode;
+    }
+  }
```

Comment:
I was actually looking into that. However the enum `SqlTypeName` isn't quite clear to me (for example: there could be multiple members mapping to struct (`Row` and `Structured`)). 

I also looked into `com.linkedin.coral.common.FuzzyUnionSqlRewriter#getUnionDataType(java.util.List<org.apache.calcite.rel.type.RelDataType>)` and looks like is also uses `isStruct` to determine struct type. Given that I would prefer keeping this way since it is even more confusing to have one if-branch plus a switch statement.